File: 2016/day5/day5.go
package main

import (
	"bufio"
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Door struct {
	UseTest   bool
	Password  string
	Password2 string
	Id        string
}

func hashMD5(input string) string {
	hash := md5.Sum([]byte(input))
	return hex.EncodeToString(hash[:])
}

func isArrayFull(arr []string) bool {
	for _, value := range arr {
		if value == "" {
			return false
		}
	}
	return true
}

func (this *Door) getInput() {
	inputFile := "input.txt"
	if this.UseTest {
		inputFile = "input-test.txt"
	}

	file, err := os.Open(inputFile)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		this.Id = line
	}
	defer file.Close()
}

func (this *Door) getPassword(isPart2 bool) string {
	if isPart2 {
		return this.Password2
	}
	return this.Password
}

func (this *Door) checkPassword() {
	password2Arr := make([]string, 8)
	num := 0
	hashed := ""
	hashStr := ""
	for len(this.Password) < 8 || !isArrayFull(password2Arr) {
		subset := ""
		for subset != "00000" && num < 100_000_000 {
			hashStr = this.Id + strconv.Itoa(num)
			hashed = hashMD5(hashStr)
			subset = hashed[:5]
			num++
		}
		if len(this.Password) < 8 {
			this.Password += hashed[5:6]
		}
		if !isArrayFull(password2Arr) {
			index, err := strconv.Atoi(hashed[5:6])
			if err != nil {
				continue
			}
			if index >= 0 && index < len(password2Arr) && password2Arr[index] == "" {
				password2Arr[index] = hashed[6:7]
			}
		}
	}
	this.Password2 = strings.Join(password2Arr, "")
}

func main() {
	door := &Door{
		UseTest: false,
	}
	door.getInput()
	door.checkPassword()
	fmt.Println("Day 5 part 1:", door.getPassword(false))
	fmt.Println("Day 5 part 2:", door.getPassword(true))
	// Total Runtime ~4.9s
}

File: 2017/day5/day5.py
from typing import List, Optional, Tuple


class CPU:
    def getInput(self) -> List[List[str]]:
        inputFile = "input-test.txt" if self.useTest else "input.txt"
        with open(inputFile, "r") as file1:
            return [int(line.strip()) for line in file1.readlines()]

    def __init__(self, useTest: Optional[bool] = False) -> None:
        self.useTest = useTest

    def getSteps(self, isPart2: Optional[bool] = False) -> int:
        self.jumpList = self.getInput()
        steps = index = 0

        while 0 <= index < len(self.jumpList):
            jump = self.jumpList[index] 
            offset = 1
            if isPart2:
                offset = -1 if jump >= 3 else 1 
            self.jumpList[index], index, = jump + offset , jump + index
            steps += 1
        return steps


if __name__ == "__main__":
    cpu = CPU()
    print("Day 5 part 1:", cpu.getSteps())
    print("Day 5 part 2:", cpu.getSteps(True))
    # Total Runtime ~7.41s
